File: language/typescript/standard-library/data-structures/graphSearch.ts
/**
 * 手写 BFS / DFS
 *
 * 考点：
 * - BFS 用队列，适合最短路径。
 * - DFS 用栈/递归，适合连通分量、拓扑。
 */

export type Graph<T> = Map<T, T[]>;

export function* bfs<T>(graph: Graph<T>, start: T): Generator<T> {
  const visited = new Set<T>();
  const queue: T[] = [start];
  visited.add(start);

  while (queue.length > 0) {
    const node = queue.shift()!;
    yield node;
    for (const neighbor of graph.get(node) ?? []) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        queue.push(neighbor);
      }
    }
  }
}

export function* dfs<T>(graph: Graph<T>, start: T): Generator<T> {
  const visited = new Set<T>();
  const stack: T[] = [start];

  while (stack.length > 0) {
    const node = stack.pop()!;
    if (visited.has(node)) continue;
    visited.add(node);
    yield node;
    const neighbors = graph.get(node) ?? [];
    for (let i = neighbors.length - 1; i >= 0; i--) {
      stack.push(neighbors[i]);
    }
  }
}

export function shortestPath<T>(graph: Graph<T>, start: T, end: T): T[] | null {
  if (start === end) return [start];
  const visited = new Set<T>([start]);
  const queue: [T, T[]][] = [[start, [start]]];

  while (queue.length > 0) {
    const [node, path] = queue.shift()!;
    for (const neighbor of graph.get(node) ?? []) {
      if (visited.has(neighbor)) continue;
      const newPath = [...path, neighbor];
      if (neighbor === end) return newPath;
      visited.add(neighbor);
      queue.push([neighbor, newPath]);
    }
  }
  return null;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const graph = new Map([
    ["A", ["B", "C"]],
    ["B", ["D"]],
    ["C", ["D"]],
    ["D", []],
  ] as [string, string[]][]);
  console.log([...bfs(graph, "A")]);
  console.log([...dfs(graph, "A")]);
  console.log(shortestPath(graph, "A", "D"));
}

File: language/python/standard-library/data_structures/graph_search.py
"""
手写 BFS / DFS。

考点：
- BFS 用队列，适合最短路径（无权图）。
- DFS 用栈（或递归），适合拓扑、连通分量。
- visited 集合避免重复访问。
"""

from __future__ import annotations

from collections import deque
from collections.abc import Callable, Iterator, Mapping
from typing import TypeVar

T = TypeVar("T")


Graph = Mapping[T, list[T]]


def bfs(graph: Graph[T], start: T) -> Iterator[T]:
    """Breadth-first traversal of ``graph`` starting from ``start``."""
    visited: set[T] = set()
    queue: deque[T] = deque([start])
    visited.add(start)

    while queue:
        node = queue.popleft()
        yield node
        for neighbor in graph.get(node, []):
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)


def dfs(graph: Graph[T], start: T) -> Iterator[T]:
    """Depth-first traversal of ``graph`` starting from ``start``."""
    visited: set[T] = set()
    stack: list[T] = [start]

    while stack:
        node = stack.pop()
        if node in visited:
            continue
        visited.add(node)
        yield node
        for neighbor in reversed(graph.get(node, [])):
            if neighbor not in visited:
                stack.append(neighbor)


def shortest_path(graph: Graph[T], start: T, end: T) -> list[T] | None:
    """Return the shortest path from ``start`` to ``end`` using BFS."""
    if start == end:
        return [start]

    visited: set[T] = {start}
    queue: deque[tuple[T, list[T]]] = deque([(start, [start])])

    while queue:
        node, path = queue.popleft()
        for neighbor in graph.get(node, []):
            if neighbor in visited:
                continue
            new_path = path + [neighbor]
            if neighbor == end:
                return new_path
            visited.add(neighbor)
            queue.append((neighbor, new_path))

    return None


if __name__ == "__main__":
    graph = {
        "A": ["B", "C"],
        "B": ["D"],
        "C": ["D"],
        "D": [],
    }
    print(list(bfs(graph, "A")))  # ['A', 'B', 'C', 'D']
    print(list(dfs(graph, "A")))  # ['A', 'B', 'D', 'C']
    print(shortest_path(graph, "A", "D"))  # ['A', 'B', 'D']
